src/main/scala/is/hail/expr/Type.scala

```diff
@@ -225,6 +225,13 @@ sealed abstract class Type extends Serializable {
   def fundamentalType: Type = this
 }
 
+case object TUnit extends Type {
+  override def toString = "Unit"
+  override def ordering(missingGreatest: Boolean): Ordering[is.hail.annotations.Annotation] = throw new UnsupportedOperationException("No ordering on Unit")
+  override def scalaClassTag: scala.reflect.ClassTag[_ <: AnyRef] = throw new UnsupportedOperationException("No ClassTag for Unit")
+  override def typeCheck(a: Any): Boolean = throw new UnsupportedOperationException("No elements of Unit")
```

Comment:
The unit type is the type with a single element.  I'm not against this, but up until now, we have  used empty struct instead of an explicit unit type.  They are synonymous.  Ordering should always be equal (extend to missing).  The class tag should be classTag[Unit].  `typeCheck` should allow `null` or `()`.  Or maybe you wanted an empty type (bottom)?

Reply:
I need an explicit unit `Type` because every IR has a `Type`.

I agree that, in theory, the unit type has these properties.

In the IR, it's an error to ever try to apply any operation (other than sequencing) to a statement, so implementing these methods removes a safety harness.

I could change it's name to `TStatement` or `TVoid`. It's just the type of something that doesn't produce a value, having only a side-effect.

Reply:
> I need an explicit unit Type because every IR has a Type.

One traditional solution is to have a 2-level IR, with statements containing exprs, and exprs only having types.

Stmt = Begin [Stmt] | Write Expr | ...

Expr = Add Type Expr Expr | ...

Another option is to make the type null for statement IR nodes.

Void is probably a better name for what you want.  You should probably also make it non-realizable.

Reply:
I'll close for now. I need to build more of the compiler to understand the tradeoffs of the different approaches.